### PR TITLE
Added a inputView option to SearchBox

### PR DESCRIPTION
**Pull request by JasonStoltz**
## Description

Adding this allows for much simpler, targeted view customization.
Instead of having to deal with all of the Downshift plumbing and logic,
this lets you very easily target just the html.

For example, if you just want to hide the search button, or inject an icon, etc. This makes it much easier. CC @scottybollinger @zumwalt, this is something you had dealt with before.

```jsx
                    <SearchBox
                      inputView={({
                        getAutocomplete,
                        getButtonProps,
                        getInputProps
                      }) => (
                        <>
                          <div className="sui-search-box__wrapper">
                            <input
                              {...getInputProps({ placeholder: "I am thor" })}
                            />
                            {getAutocomplete()}
                          </div>
                          <input {...getButtonProps()} />
                        </>
                      )}
```

See the ADVANCED.md updates for some more examples.

## List of changes

- Added a inputView option to SearchBox

## Associated Github Issues

#163

### CODE REVIEW

File: packages/react-search-ui-views/src/SearchBox.js [hunk 1]
```diff
@@ -53,25 +56,35 @@ function SearchBox(props) {
                 appendClassName("sui-search-box", className) + autocompleteClass
               }
             >
-              <div className="sui-search-box__wrapper">
-                <input
-                  {...getInputProps({
+              <InputView
+                getInputProps={additionalProps =>
```
Comment (constancecchen) [inline packages/react-search-ui-views/src/SearchBox.js]: I don't think I'm understanding where `additionalProps` are coming from or how devs should pass this in in a final usage situation. Is there any chance we could write an extra test that shows a custom `inputView` function with additional props for both the search input and the button input?

File: packages/react-search-ui-views/src/SearchBox.js [hunk 2]
```diff
@@ -53,25 +56,35 @@ function SearchBox(props) {
...
+                })}
+                getAutocomplete={() => {
+                  if (
+                    useAutocomplete &&
+                    isOpen &&
+                    allAutocompletedItemsCount > 0
+                  ) {
+                    return <AutocompleteView {...props} {...downshiftProps} />;
+                  } else {
+                    return null;
```
Comment (constancecchen) [inline packages/react-search-ui-views/src/SearchBox.js]: Very minor code style suggestion - but you might get fewer lines/a smoother flow by turning this into a ternary: 

```jsx
getAutocomplete={() => {
  const showAutocomplete =
    (useAutocomplete && isOpen && allAutocompletedItemsCount > 0);
  return showAutocomplete
    ? <AutocompleteView {...props} {...downshiftProps} />
    : null;
}}
```

File: packages/react-search-ui-views/src/SearchBox.js (hunk 1 above)
Comment (constancecchen) [reply packages/react-search-ui-views/src/SearchBox.js]: Ah, duh, I see an example in the Storybook file below. I'd still really like to see a test written for the new functionality possible, but totally get if not or if it's better suited for an integration test.

I'm also a little curious about the purpose of the markdown file examples vs the storybook examples - would it make sense to combine them at one point so all examples are in one location? Or are they specifically split up for a reason?

File: ADVANCED.md [hunk 3]
```diff
@@ -295,6 +295,62 @@ The input from `SearchBox` will be used to trigger a new search query. That quer
...
+```
+
+Or you can also just customize the autocomplete dropdown:
+
+```jsx
+<SearchBox
+  autocompleteView={({ autocompletedResults, getItemProps }) => (
+    <div className="sui-search-box__autocomplete-container">
+      {autocompletedResults.map((result, i) => (
+        // eslint-disable-next-line react/jsx-key
```
[the remark this answers is not in the document]
Comment (JasonStoltz) [reply ADVANCED.md]: Yes, will remove it.

File: packages/react-search-ui-views/src/SearchInput.js [hunk 4]
```diff
@@ -0,0 +1,22 @@
+import PropTypes from "prop-types";
+import React from "react";
+
+function SearchInput({ getAutocomplete, getButtonProps, getInputProps }) {
```
[the remark this answers is not in the document]
Comment (JasonStoltz) [reply packages/react-search-ui-views/src/SearchInput.js]: I don't have a particular reason for doing it one way or the other.

File: packages/react-search-ui-views/src/SearchBox.js (hunk 2 above)
Comment (JasonStoltz) [reply packages/react-search-ui-views/src/SearchBox.js]: Sometimes the explicit if / else is just easier to read. I don't mind changing it though.

Comment (constancecchen) [reply packages/react-search-ui-views/src/SearchBox.js]: I'm good either way!

File: packages/react-search-ui-views/src/SearchInput.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (constancecchen) [reply packages/react-search-ui-views/src/SearchInput.js]: Cool, no worries! I'll keep this format in mind if I ever contribute/write a component! 🤞

[the remark this answers is not in the document]
Comment (JasonStoltz) [reply packages/react-search-ui-views/src/SearchInput.js]: @constancecchen I am lax on code style. I have probably used both styles for functional components.

File: packages/react-search-ui-views/src/SearchBox.js (hunk 1 above)
Comment (JasonStoltz) [reply packages/react-search-ui-views/src/SearchBox.js]: The markdown files are the documentation, meant for people using the project. The storybook is more for us, being able to test and styles components in various states in isolation. I would very much like to consolidate this. I think, ideally, that component reference section could just be replaced with storybook.

Comment (JasonStoltz) [reply packages/react-search-ui-views/src/SearchBox.js]: @constancecchen I didn't realize that we don't have a test for this. It was in severe need. I added tests that cover both `getInputProps` and `getButtonProps`. I also did some reworking so they preserve css class names, which would have been a big issue. Thank you for your keen eyes.

I unfortunately could not find a way to test `getAutocomplete`, but I need to move on from this PR. The smoke tests in storybook and the sandbox will have to be enough for now.

I also added notes about this pattern to the README. I'd like to expand more views, in the future, to use this pattern: https://github.com/downshift-js/downshift#prop-getters

**Comment by JasonStoltz** [comment]
@constancecchen Ready for another look!

File: packages/react-search-ui-views/src/SearchBox.js [hunk 5]
```diff
@@ -57,23 +57,31 @@ function SearchBox(props) {
               }
             >
               <InputView
-                getInputProps={additionalProps =>
-                  getInputProps({
+                getInputProps={additionalProps => {
+                  const { className, ...rest } = additionalProps || {};
```
Comment (constancecchen) [inline packages/react-search-ui-views/src/SearchBox.js]: Awesome catch here! 👏

File: packages/react-search-ui-views/src/__tests__/SearchBox.test.js [hunk 6]
```diff
@@ -60,7 +60,95 @@ it("applies className from inputProps to input element", () => {
...
+      .shallow()
+      .find("input")
+      .at(0);
+
+    button = wrapper
+      .dive("Downshift")
+      .find("inputView")
+      .shallow()
+      .find("input")
+      .at(1);
```
Comment (constancecchen) [inline packages/react-search-ui-views/src/__tests__/SearchBox.test.js]: Good ol' Jest chaining + prettier autolinting :) This is a really clever way of DRYing out the various child diving/shallowing, huge shout out! 💯 

Optional suggestion / feel free to disregard this/ship as-is (like you mentioned, probably good to move on at this point), but if you wanted to be even DRYer and possibly sidestep prettier's linting you could try something like this:

```js
inputView = wrapper
  .dive("Downshift")
  .find("inputView")
  .shallow(); // I think Prettier still multi-line this, oh well
input = inputView.find('input').at(0);
button = inputView.find('button').at(1);
```
and then down on line 113 you can reuse inputView again:
```js
expect(inputView.find('.some_custom_wrapper_class')).toHaveLength(1);
```

File: packages/react-search-ui-views/src/SearchBox.js (hunk 1 above)
Comment (constancecchen) [reply packages/react-search-ui-views/src/SearchBox.js]: All good, this is an amazing amount of extra tests/documentation!! Thanks so much for going above and beyond! 😁

File: ADVANCED.md [hunk 7]
```diff
@@ -314,21 +314,38 @@ customizations, like simply hiding the search button, this is often overkill.
 ```
 
 You can also just customize the input section of the search box. Useful for things
-like hiding the submit button:
+like hiding the submit button or rearranging dom structure:
```
Comment (constancecchen) [inline ADVANCED.md]: Super minor grammar nit: DOM should probably be capitalized here / for consistency (it's capitalized elsewhere on the page)

Comment (JasonStoltz) [reply ADVANCED.md]: Good point! Sorry, hit the merge button.

File: packages/react-search-ui-views/src/__tests__/SearchBox.test.js (hunk 6 above)
Comment (JasonStoltz) [reply packages/react-search-ui-views/src/__tests__/SearchBox.test.js]: Good point! Sorry, hit the merge button.

**Pull request merged**